Detect and report BIOS configuration drift on deprov

ScottGarman opened the pull request:
This is work I've been doing in another PR that I believe is ready for final review and would be low-risk to deploy to production, given that only BIOS config drift is detected and configs are not yet applied.

ScottGarman committed:
Detect and report BIOS vendor/version on deprov

This currently works for Dell and Supermicro servers, and can be
expanded for additional vendors as needed.

Signed-off-by: Scott Garman <sgarman@packet.com>

ScottGarman committed:
Download our BIOS config files during deprov

The bios-configs-latest tarball includes our BIOS config files and a
manifest.txt file which provides a lookup table to map a plan name and
BIOS vendor to a configuration filename. The manifest also has an
enforcement status field which can be set to "testing" if we want to
only check and report BIOS drift but not apply the config.

Signed-off-by: Scott Garman <sgarman@packet.com>

ScottGarman committed:
Save current BIOS config to a local file on deprov

This will be used to compare the server's settings to what we're
expecting them to be.

Signed-off-by: Scott Garman <sgarman@packet.com>

ScottGarman committed:
Detect and report BIOS configuration drift on deprov

These functions compare the current BIOS configuration to what we're
expecting from the configs in our firmware repo. This isn't as simple as
dumping the config and performing a diff - the configs must be
normalized by filtering out irrelevant config sections that might change
from server to server, but don't represent meaningful drift.

The filtering approach used here is admittedly messy and not very
maintainable over the long term, but this effort is being done as a
stopgap until a separate, microservice-based approach to maintaining
firmware and BIOS settings is completed.

Signed-off-by: Scott Garman <sgarman@packet.com>

ScottGarman committed:
Remove redundant "Error during" autofail stage message

Signed-off-by: Scott Garman <sgarman@packet.com>

mikemrm commented on docker/scripts/functions.sh:
```diff
@@ -98,6 +98,47 @@ function posttink() {
+	vendor=unknown
+
+	# Check for Dell
+	dell_output=$(/opt/dell/srvadmin/bin/idracadm7 get BIOS.SysInformation 2>&1)
```
Could we potentially use the exit code here instead of grepping the output?

ScottGarman replied:
I think so. Let me investigate.

ScottGarman replied:
Yep, this works well and is significantly cleaner. Thanks for the suggestion!

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -108,6 +198,128 @@ function dns_resolvers() {
+			del(.SystemConfiguration.Components[] | select(.FQDD=="BIOS.Setup.1-1").Attributes[] | select(.Name=="SetBootOrderEn")) |
+			del(.SystemConfiguration.Components[] | select(.FQDD=="BIOS.Setup.1-1").Attributes[] | select(.Name=="BiosBootSeq")) |
+			del(.SystemConfiguration.Components[] | select(.FQDD=="iDRAC.Embedded.1").Attributes[] | select(.Name=="NIC.1#DNSRacName")) |
+			del(.SystemConfiguration.Components[] | select(.FQDD=="System.Embedded.1").Attributes[] | select(.Name=="ServerOS.1#HostName"))' \
```
would this be easier with `gron | grep -v`?

ScottGarman replied:
Great idea, I'm looking into adding this tonight.

ScottGarman replied:
I've added the gron binary to OSIE, will hold off on refactoring to use it just yet as we discussed earlier.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -98,6 +98,43 @@ function posttink() {
 
+# returns a string of the BIOS vendor: "dell", "supermicro", or "unknown"
+function detect_bios_vendor() {
+	set +e
```
why does this need to be done? It looks like you handle any error paths so shouldn't need to do this.

ScottGarman replied:
This is because we're potentially running the dell/supermicro bios utilities on platforms which don't support them, so I'm trying to prevent the script from exiting/triggering autofail due to the non-zero exit status those commands will return.

mmlb replied:
Thats ok though, you won't have bash exit because you handle all the error cases:
```
[manny@dellnix:~]$ bash <<EOF
set -e
if /opt/dell/srvadmin/bin/idracadm7; then                           
  echo we should not see this, there is no idracadm7
elif /opt/supermicro/sum/sum; then
  echo we should not see this, there is no sum tool
else
  echo ok so everything errored, its ok bash does not die because we handled it with if and else
fi
EOF
bash: line 2: /opt/dell/srvadmin/bin/idracadm7: No such file or directory
bash: line 4: /opt/supermicro/sum/sum: No such file or directory
ok so everything errored, its ok bash does not die because we handled it with if and else
```
This shouldn't trigger the exiting/autofail either.

ScottGarman replied:
Ah, it is now since I'm no longer using intermediate variables to store/parse the output of the commands, which was my previous approach. Good catch.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -98,6 +98,43 @@ function posttink() {
+	vendor=unknown
+
+	# Check for Dell
+	if /opt/dell/srvadmin/bin/idracadm7 get BIOS.SysInformation >/dev/null 2>&1; then
```
this can just be `&>/dev/null` which I think we use around here too

ScottGarman replied:
Fixed.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -98,6 +98,43 @@ function posttink() {
+		version=$(echo "${dell_output}" | awk -F "=" '/^#SystemBiosVersion/ {print $2}')
+	fi
+	if [[ "${vendor}" == "Supermicro" ]]; then
+		sum_output=$(/opt/supermicro/sum/sum -c GetDmiInfo)
```
I'd do away with the `*_output` variables (which aren't being set a `local` so are polluting the global space fyi)

ScottGarman replied:
Fixed.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -98,6 +98,43 @@ function posttink() {
+# returns a string of the BIOS vendor: "dell", "supermicro", or "unknown"
+function detect_bios_vendor() {
+	set +e
+	vendor=unknown
```
should be `local`

ScottGarman replied:
Fixed.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -98,6 +98,43 @@ function posttink() {
+function detect_bios_version() {
+	local vendor=$1
+
+	version=unknown
```
should be `local`

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -135,6 +135,59 @@ function detect_bios_version() {
+		echo "$images_ip        github.com"
+		echo "$images_ip        github-mirror.packet.net"
+	} >>/etc/hosts.new
+	cp -f /etc/hosts.new /etc/hosts
```
can you make this into its own function so we can share with osie's fetch-image step please.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -135,6 +135,59 @@ function detect_bios_version() {
+
+	echo "Downloading latest BIOS configurations"
+	curl https://github-mirror.packet.net/downloads/bios-configs-latest.tar.gz --output bios-configs-latest.tar.gz
+	curl https://github-mirror.packet.net/downloads/bios-configs-latest.tar.gz.sha256 --output bios-configs-latest.tar.gz.sha256
```
strongly encourage a `-f` here, since of we get an "http error" code curl will still return 0 and now we're running with garbage input. `-f` makes curl fail on those 4xx/5xx status.

ScottGarman replied:
Added (in long form of course, --fail).

mmlb replied:
of course! :+1:

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -135,6 +135,59 @@ function detect_bios_version() {
+		echo "Error: missing BIOS config manifest file"
+		return 1
+	fi
+	status=$(grep "^${plan} \+${vendor}" bios-configs-latest/manifest.txt | awk '{print $4}' || true)
```
should have `local`

ScottGarman replied:
Fixed.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -135,6 +135,59 @@ function detect_bios_version() {
+		echo "Error: missing BIOS config manifest file"
+		return 1
+	fi
+	configfile=$(grep "^${plan} \+${vendor}" bios-configs-latest/manifest.txt | awk '{print $3}' || true)
```
should be `local`

ScottGarman replied:
Fixed.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -198,6 +198,40 @@ function dns_resolvers() {
+	local plan=$2
+
+	# Check for a BIOS config for this plan
+	config_file=$(lookup_bios_config "${plan}" "${vendor}")
```
should be `local`

ScottGarman replied:
Fixed.

mmlb commented on docker/scripts/functions.sh:
```diff
@@ -210,6 +210,82 @@ function retrieve_current_bios_config() {
+	local config_file="bios-configs-latest/$2"
+	local config_file_normalized="${config_file}.normalized"
+	local current_config="current_bios.txt"
+	local current_config_normalized="${current_config}.normalized"
```
vendor is unused right? Also why not just pass the normalized file paths to the func?

ScottGarman replied:
Yep, vendor was an artifact from before I broke out the vendor-specific normalization code into its own functions. And I've specified the bios-configs-latest/ file path in the function call rather than inside the function.

ScottGarman committed:
Move IPN cache DNS configuration to its own function

This will allow it to be reused in the deprovision process for BIOS
config tarball downloads.

Signed-off-by: Scott Garman <sgarman@packet.com>

ScottGarman committed:
Add gron (grep for JSON) to OSIE

This is a handy alternative to jq in certain situations, and will be
used to filter/normalize BIOS config files.

Signed-off-by: Scott Garman <sgarman@packet.com>
